qiita_db/sql_connection.py

```diff
@@ -633,3 +581,276 @@ def get_temp_queue(self):
         self.create_queue(temp_queue_name)
 
         return temp_queue_name
+
+
+def _checker(func):
+    """Decorator to check that methods are executed inside the context"""
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if not self._is_inside_context:
+            raise RuntimeError(
+                "Operation not permitted. Transaction methods can only be "
+                "invoked within the context manager.")
+        return func(self, *args, **kwargs)
+    return wrapper
+
+
+class Transaction(object):
+    """A context manager that encapsulates a DB transaction
+
+    A transaction is defined by a series of consecutive queries that need to
+    be applied to the database as a single block.
+
+    Parameters
+    ----------
+    name : str
+        Name of the transaction.
+
+    Notes
+    -----
+    When the execution leaves the context manager, any remaining queries in
+    the transaction will be executed and committed.
+    The Transaction methods can only be executed inside a context
+    """
+
+    _regex = re.compile("^{(\d+):(\d+):(\d+)}$")
+
+    def __init__(self, name):
+        # The name is useful for debugging, since we can identify the
+        # failed queue in errors
+        self._name = name
+        self._queries = []
+        self._results = []
+        self.index = 0
+        self._conn_handler = SQLConnectionHandler()
+        self._is_inside_context = False
+
+    def __enter__(self):
+        self._is_inside_context = True
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        # We need to wrap the entire function in a try/finally because
+        # at the end of the function we need to set _is_inside_context to false
+        try:
+            status = self._conn_handler._connection.get_transaction_status()
+            if exc_type is not None:
+                # An exception occurred during the execution of the transaction
+                # Make sure that we leave the DB w/o any modification
+                self.rollback()
+            elif self._queries:
+                # There are still queries to be executed, execute them
+                # It is safe to use the execute method here, as internally is
+                # wrapped in a tr/except and rollbacks in case of failure
+                self.execute()
+            elif status != TRANSACTION_STATUS_IDLE:
+                # There are no queries to be executed, however, the transaction
+                # is still not committed. Commit it so the changes are not lost
+                self.commit()
+        finally:
+            self._is_inside_context = False
+
```

Comment:
on exit, can the connection be explicitly closed? or is there a benefit in it possibly staying open? just concerned as the destructors with psycopg2 haven't always worked implicitly as expected